### DISCUSSION TITLE
DataPortal extension source generator

### DISCUSSION THREAD

**Discussion by StefanOssendorf** [MarimerLLC/csla#3681] [General]
Hello everyone.
I'd like to promote my new source generator for DataPortal extensions :)  
You can find it here: [DataPortal extension generator repo](https://github.com/StefanOssendorf/Csla.DataPortalExtensions)

It's a source gen which generates for each `Fetch`, `Create`, `Execute` and `Delete` method a compile and type safe extension method.
A short example would be:
```csharp

// Your CSLA business object.
public class Foo : BusinessBase<Foo> {
    [Fetch]
    private async Task Fetch(string a, int b, int? c, Guid id) {
        // fetch logic
    }
	
    [Create]
    private void Create(string name, string address) {
        // create logic
    }
	
    [Delete]
    private async Task ById(Guid id) {
        // delete logic
    }
}

// Class which is used to generate the extension methods into
[Ossendorf.Csla.DataPortalExtensionsGenerator.DataPortalExtensions] // Attribute which says generate the extensions into this type
public static partial class DataPortalExtensions {
}
```

The source generator will generate the following code (I removed some stuff for type clarification for easiere reading):
```csharp
static partial class DataPortalExtensions {
	public static Task<Foo> Fetch(this IDataPortal<Foo> portal, string a, int b, int? c, Guid id) => portal.FetchAsync(a, b, c, id);
	public static Task<Foo> Create(this IDataPortal<Foo> portal, string name, string address) => portal.CreateAsync(name, address);
	public static Task ById(this IDataPortal<Foo> portal, Guid id) => portal.DeleteAsync(id);
}
```

If you have any question feel free to open issues and ask your qestion :)

Have a nice day
Stefan

**Comment by Chicagoan2016** [comment]
I love this conversation, I was going to start a discussion about code generation and Csla. In our team, we could  use code generation in a number of places, we use stored procs and ADO.NET so data access could be generated entirely. Same goes for Csla business object properties and Data portal methods. Hopefully our developers will only write code that implements 'meaty' processes such as business rules, authorization rules etc.
